nn/nnclient.py

```diff
@@ -0,0 +1,305 @@
+#!/usr/bin/env python3
+"""
+    Reynir: Natural language processing for Icelandic
+
+    Neural Network Query Client
+
+    Copyright (C) 2018 Miðeind ehf.
+
+       This program is free software: you can redistribute it and/or modify
+       it under the terms of the GNU General Public License as published by
+       the Free Software Foundation, either version 3 of the License, or
+       (at your option) any later version.
+       This program is distributed in the hope that it will be useful,
+       but WITHOUT ANY WARRANTY; without even the implied warranty of
+       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+       GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see http://www.gnu.org/licenses/.
+
+
+    This module implements a client that connects to a middleware
+    neural network server (see nnserver/nnserver.py), which in turn
+    connects to a TensorFlow model server.
+
+"""
+
+import json
+import requests
+import logging
+
+from settings import Settings
+import tokenizer
+
+from nn import nntree
+from nn.nntree import ParseResult
+from reynir.binparser import BIN_Token
+from reynir import bintokenizer
+
+
+def index_text(text):
+    """ Segments contiguous (Icelandic) text into paragraphs and sentences
+        and returns:
+            dictionary of sentence indices to sentences
+            dictionary of paragraph index to constituent sentence indices"""
+    text = "[[ " + " ]] [[ ".join(text.split("\n")) + " ]]"
+    tok_stream = bintokenizer.tokenize(text)
+
+    pgs = tokenizer.paragraphs(tok_stream)
+    pg_idx_to_sent_idx = dict()
+    sent_idx_to_sent = dict()
+    curr_sent_idx = 0
+    curr_pg_idx = 0
+
+    for pg in pgs:
+        sent_idxs = []
+        for (idx, sent) in pg:
+            curr_sent = list(filter(BIN_Token.is_understood, sent))
+            curr_sent = " ".join([tok.txt for tok in curr_sent])
```

Comment:
Square brackets are unnecessary; for safety it might be better to say ``" ".join(tok.txt for tok in curr_sent if tok.txt)``

Reply:
...and, even better, to define a function:
```
 def to_txt(t): return tok.val[1] if tok.kind == TOK.PUNCTUATION else tok.txt
```
and use that one to extract the token text (this fetches normalized punctuation, for instance Icelandic double quotes instead of English ones, regular hyphens for en- and em-dashes, etc.)

Reply:
Such a function (`normalized_text()` and `normalized_text_from_tokens()` is now available from the Tokenizer package).